### ISSUE TITLE
Oracledb module silently crashes

### ISSUE THREAD

**Issue by Tomkinen** [oracle/node-oracledb#607]
I am running nodejs (version 4.7.2) in windows server 2012 R2. Oracledb is version 1.12.2. Compiler is visual studio 2013.

When running nodejs application that uses oracledb module, the application randomly crashes without any error messages visible in command prompt.

Usually the nodejs application crashes in matter of seconds after starting to use it. Especially if the query made by oracledb would return big amount of data.

I had to install segfault-handler -module to be able to get some errors logged.

Here is what segfault-handler is returning as error message. It seems that the binaries oracledb is using are causing the error:

PID 2528 received SIGSEGV for address: 0x7a91ed78
c:\nodejs_app\indication\node_modules\segfault-handler\src\stackwalker.cpp (941): StackWalker::ShowCallstack
c:\nodejs_app\indication\node_modules\segfault-handler\src\segfault-handler.cpp (116): segfault_handler
00007FFC8AE05F32 (ntdll): (filename not available): RtlRestoreContext
00007FFC8AE04753 (ntdll): (filename not available): RtlRaiseException
00007FFC8AE4309A (ntdll): (filename not available): KiUserExceptionDispatcher
c:\nodejs_app\indication\node_modules\oracledb\src\njs\src\njsconnection.h (457): eBaton::~eBaton
c:\nodejs_app\indication\node_modules\oracledb\src\njs\src\njsconnection.cpp (3796): Connection::Async_AfterExecute
00007FF672F9A4C3 (node): (filename not available): uv_fs_poll_getpath
00007FF672F9A427 (node): (filename not available): uv_fs_poll_getpath
00007FF672F8876B (node): (filename not available): uv_async_send
00007FF672F85284 (node): (filename not available): BIO_asn1_get_suffix
00007FF672F85BB8 (node): (filename not available): uv_run
00007FF672DFF480 (node): (filename not available): node::CreateEnvironment
00007FF672DFF776 (node): (filename not available): node::Start
00007FF6734F0EA4 (node): (filename not available): lh_node_usage_stats_bio
00007FFC8AC713D2 (KERNEL32): (filename not available): BaseThreadInitThunk
00007FFC8ADC5444 (ntdll): (filename not available): RtlUserThreadStart

Files "oracledb\src\njs\src\njsconnection.h" and "oracledb\src\njs\src\njsconnection.cpp" are causing the crash. Should i install older version of oracledb as that seems to work ok? I have used oracledb for 2 years without this error ever happening. Errors started to happen after updating nodefrom 4.2.6 to 4.7.2 and oracledb from 1.6.0 to 1.12.2.

I will give more information if you require. Would be very nice to fix this properly (without need to use older version).

**Comment by cjbj** [comment]
Sorry about the crash.  In order for us to diagnose anything, you'll need to narrow down what triggers it.

**Comment by Tomkinen** [comment]
Crash is triggered in basically any code that connects to db, like for example connection to "table"-table: "connection.execute("select * from table",[],{outFormat:oracledb.OBJECT},function(err,result){".

For your information i reinstalled oracledb 1.6.0 and everything works ok again. ...Still would prefer to use latest version without errors.

**Comment by cjbj** [comment]
Re-rebuild?  Try upgrading only one of Node or node-oracledb at a time?  Upgrade to the latest Node - 4.7.3? Try Node 6? Remove any library clashes in PATH if you have multiple Oracle software installs.

**Comment by Tomkinen** [comment]
Yes. You are right. I should try to re-rebuild to see if it has impact. I will get back once i have done that.

**Comment by yancald** [comment]
Hi, I have similar issue.  it breaks at same place in "oracledb\src\njs\src\njsconnection.h" and "oracledb\src\njs\src\njsconnection.cpp". I find that the pattern is:
1. set "maxRows" to >1000
2. There are more than 130 columns in "select "  statement.

**Comment by Tomkinen** [comment]
@yancald Exactly! Thanks for testing the scenarios which cause this! Appreciate it. Same here. Reinstalling does not help to solve this, so the label of "Install & Configuration" might not be accurate.

I have maxRows set to 10000 for these cases which cause the error. Select should return thousands of rows (not exceeding 10000 in any case). Less than 130 columns in select. So the pattern number 1 is accurate in my case.

**Comment by cjbj** [comment]
Please give some more details: schema, code etc.  The original description "Crash is triggered in basically any code that connects to db" case is very broad unlike the specific case @yancald describes.  So what are the causes?  Are you streaming (which seems to be a common factor in #609)?

@Tomkinen I would suggest using an ResultSet or streaming instead of setting such a big maxRows: without those you will be allocating huge chunks of memory that you many not be using, which is very wasteful.

**Comment by Tomkinen** [comment]
@cjbj I have never used ResultSet or streaming with oracledb. All queries have been done with connection.execute with outFormat set to be oracledb.OBJECT. Most queries have reasonable small maxRows, but this particular application which crashes can get thousands of rows from db with single query. Therefore i should any way start to use ResultSet, no matter would the application crash or not. I will have to do some tweaks to my backend API code to utilize ResultSet, do some testing again and then get back to you. Thanks for the help so far.

**Comment by cjbj** [comment]
@Tomkinen @yancald I can't reproduce this on OS X and it hasn't appeared in our stress tests. Got a standalone testcase you can share?

**Comment by cjbj** [comment]
@Tomkinen Thanks for the update.  I tried thousands of rows and hundreds of columns. But no doubt a testcase from you would let me zero in on the problem directly.

I am not a Windows user :)

**Comment by sant123** [comment]
I am running nodejs (version 7.6.0) in windows 10. Oracledb is version 1.12.2. Compiler is visual studio 2015.

I get the following error when I fetch a clob as string.

``` javascript
oracledb.fetchAsString = [oracledb.CLOB];
```

PID 5424 received SIGSEGV for address: 0x1ab40229
SymInit: Symbol-SearchPath: '.;D:\santiagoah\informacion\Desktop\exportacion-oracle;C:\Program Files\nodejs;C:\Windows;C:\Windows\system32;SRV*C:\websymbols*http://msdl.microsoft.com/download/symbols;', symOptions: 530, UserName: 'SantiagoAH'
OS-Version: 10.0.14393 () 0x100-0x1
d:\santiagoah\informacion\desktop\exportacion-oracle\node_modules\segfault-handler\src\stackwalker.cpp (941): StackWalker::ShowCallstack
d:\santiagoah\informacion\desktop\exportacion-oracle\node_modules\segfault-handler\src\segfault-handler.cpp (116): segfault_handler
00007FF9395ABD00 (ntdll): (filename not available): RtlInitializeCriticalSection
00007FF93958960B (ntdll): (filename not available): RtlImageNtHeaderEx
00007FF9395E9CBA (ntdll): (filename not available): KiUserExceptionDispatcher
d:\santiagoah\informacion\desktop\exportacion-oracle\node_modules\oracledb\src\njs\src\njsconnection.cpp (5605): Connection::cbDynDefine
d:\santiagoah\informacion\desktop\exportacion-oracle\node_modules\oracledb\src\dpi\src\dpistmtimpl.cpp (757): StmtImpl::defineCallback
00007FF8D7AD0FFE (OraOCIEI12): (filename not available): kpugc
00007FF8D6B173AD (OraOCIEI12): (filename not available): kpcdasg
00007FF8D6AD0BE7 (OraOCIEI12): (filename not available): sigpidu
00007FF8D6AA20DC (OraOCIEI12): (filename not available): sigpidu
00007FF8D84D5D1C (OraOCIEI12): (filename not available): nsrecv
00007FF8D68A6557 (OraOCIEI12): (filename not available): upicui2
00007FF8D7AD9585 (OraOCIEI12): (filename not available): upirtr
00007FF8D7BB5E9E (OraOCIEI12): (filename not available): kpurcsc
00007FF8D7AD2A15 (OraOCIEI12): (filename not available): kpugc
00007FF8D7AD4AE6 (OraOCIEI12): (filename not available): kpugc
00007FF8D793E0CD (OraOCIEI12): (filename not available): OCIStmtFetch2
00007FF91AA72F0F (OCI): (filename not available): OCIStmtFetch2
d:\santiagoah\informacion\desktop\exportacion-oracle\node_modules\oracledb\src\njs\src\njsconnection.cpp (3367): Connection::DoFetch
d:\santiagoah\informacion\desktop\exportacion-oracle\node_modules\oracledb\src\njs\src\njsresultset.cpp (461): ResultSet::Async_GetRows
00007FF7A3E7C887 (node): (filename not available): uv_fs_poll_getpath
00007FF7A3E6B7A4 (node): (filename not available): uv_once
00007FF7A46B9DCD (node): (filename not available): uncompress
00007FF9390D8364 (KERNEL32): (filename not available): BaseThreadInitThunk
00007FF9395A70D1 (ntdll): (filename not available): RtlUserThreadStart
PID 14124 received SIGSEGV for address: 0x1ab40229
SymInit: Symbol-SearchPath: '.;D:\santiagoah\informacion\Desktop\exportacion-oracle;C:\Program Files\nodejs;C:\Windows;C:\Windows\system32;SRV*C:\websymbols*http://msdl.microsoft.com/download/symbols;', symOptions: 530, UserName: 'SantiagoAH'
OS-Version: 10.0.14393 () 0x100-0x1
d:\santiagoah\informacion\desktop\exportacion-oracle\node_modules\segfault-handler\src\stackwalker.cpp (941): StackWalker::ShowCallstack
d:\santiagoah\informacion\desktop\exportacion-oracle\node_modules\segfault-handler\src\segfault-handler.cpp (116): segfault_handler
00007FF9395ABD00 (ntdll): (filename not available): RtlInitializeCriticalSection
00007FF93958960B (ntdll): (filename not available): RtlImageNtHeaderEx
00007FF9395E9CBA (ntdll): (filename not available): KiUserExceptionDispatcher
d:\santiagoah\informacion\desktop\exportacion-oracle\node_modules\oracledb\src\njs\src\njsconnection.cpp (5605): Connection::cbDynDefine
d:\santiagoah\informacion\desktop\exportacion-oracle\node_modules\oracledb\src\dpi\src\dpistmtimpl.cpp (757): StmtImpl::defineCallback
00007FF8D7AD0FFE (OraOCIEI12): (filename not available): kpugc
00007FF8D6B173AD (OraOCIEI12): (filename not available): kpcdasg
00007FF8D6AD0BE7 (OraOCIEI12): (filename not available): sigpidu
00007FF8D6AA20DC (OraOCIEI12): (filename not available): sigpidu
00007FF8D84D5D1C (OraOCIEI12): (filename not available): nsrecv
00007FF8D68A6557 (OraOCIEI12): (filename not available): upicui2
00007FF8D7AD9585 (OraOCIEI12): (filename not available): upirtr
00007FF8D7BB5E9E (OraOCIEI12): (filename not available): kpurcsc
00007FF8D7AD2A15 (OraOCIEI12): (filename not available): kpugc
00007FF8D7AD4AE6 (OraOCIEI12): (filename not available): kpugc
00007FF8D793E0CD (OraOCIEI12): (filename not available): OCIStmtFetch2
00007FF91AA72F0F (OCI): (filename not available): OCIStmtFetch2
d:\santiagoah\informacion\desktop\exportacion-oracle\node_modules\oracledb\src\njs\src\njsconnection.cpp (3367): Connection::DoFetch
d:\santiagoah\informacion\desktop\exportacion-oracle\node_modules\oracledb\src\njs\src\njsresultset.cpp (461): ResultSet::Async_GetRows
00007FF7A3E7C887 (node): (filename not available): uv_fs_poll_getpath
00007FF7A3E6B7A4 (node): (filename not available): uv_once
00007FF7A46B9DCD (node): (filename not available): uncompress
00007FF9390D8364 (KERNEL32): (filename not available): BaseThreadInitThunk
00007FF9395A70D1 (ntdll): (filename not available): RtlUserThreadStart
PID 18112 received SIGSEGV for address: 0x1ac30229
SymInit: Symbol-SearchPath: '.;D:\santiagoah\informacion\Desktop\exportacion-oracle;C:\Program Files\nodejs;C:\Windows;C:\Windows\system32;SRV*C:\websymbols*http://msdl.microsoft.com/download/symbols;', symOptions: 530, UserName: 'SantiagoAH'
OS-Version: 10.0.14393 () 0x100-0x1
d:\santiagoah\informacion\desktop\exportacion-oracle\node_modules\segfault-handler\src\stackwalker.cpp (941): StackWalker::ShowCallstack
d:\santiagoah\informacion\desktop\exportacion-oracle\node_modules\segfault-handler\src\segfault-handler.cpp (116): segfault_handler
00007FF9395ABD00 (ntdll): (filename not available): RtlInitializeCriticalSection
00007FF93958960B (ntdll): (filename not available): RtlImageNtHeaderEx
00007FF9395E9CBA (ntdll): (filename not available): KiUserExceptionDispatcher
d:\santiagoah\informacion\desktop\exportacion-oracle\node_modules\oracledb\src\njs\src\njsconnection.cpp (5605): Connection::cbDynDefine
d:\santiagoah\informacion\desktop\exportacion-oracle\node_modules\oracledb\src\dpi\src\dpistmtimpl.cpp (757): StmtImpl::defineCallback
00007FF8D7AD0FFE (OraOCIEI12): (filename not available): kpugc
00007FF8D6B173AD (OraOCIEI12): (filename not available): kpcdasg
00007FF8D6AD0BE7 (OraOCIEI12): (filename not available): sigpidu
00007FF8D6AA20DC (OraOCIEI12): (filename not available): sigpidu
00007FF8D84D5D1C (OraOCIEI12): (filename not available): nsrecv
00007FF8D68A6557 (OraOCIEI12): (filename not available): upicui2
00007FF8D7AD9585 (OraOCIEI12): (filename not available): upirtr
00007FF8D7BB5E9E (OraOCIEI12): (filename not available): kpurcsc
00007FF8D7AD2A15 (OraOCIEI12): (filename not available): kpugc
00007FF8D7AD4AE6 (OraOCIEI12): (filename not available): kpugc
00007FF8D793E0CD (OraOCIEI12): (filename not available): OCIStmtFetch2
00007FF91AB62F0F (OCI): (filename not available): OCIStmtFetch2
d:\santiagoah\informacion\desktop\exportacion-oracle\node_modules\oracledb\src\njs\src\njsconnection.cpp (3367): Connection::DoFetch
d:\santiagoah\informacion\desktop\exportacion-oracle\node_modules\oracledb\src\njs\src\njsresultset.cpp (461): ResultSet::Async_GetRows
00007FF7A3E7C887 (node): (filename not available): uv_fs_poll_getpath
00007FF7A3E6B7A4 (node): (filename not available): uv_once
00007FF7A46B9DCD (node): (filename not available): uncompress
00007FF9390D8364 (KERNEL32): (filename not available): BaseThreadInitThunk
00007FF9395A70D1 (ntdll): (filename not available): RtlUserThreadStart
PID 16776 received SIGSEGV for address: 0x1a9b0229
SymInit: Symbol-SearchPath: '.;D:\santiagoah\informacion\Desktop\exportacion-oracle;C:\Program Files\nodejs;C:\Windows;C:\Windows\system32;SRV*C:\websymbols*http://msdl.microsoft.com/download/symbols;', symOptions: 530, UserName: 'SantiagoAH'
OS-Version: 10.0.14393 () 0x100-0x1
d:\santiagoah\informacion\desktop\exportacion-oracle\node_modules\segfault-handler\src\stackwalker.cpp (941): StackWalker::ShowCallstack
d:\santiagoah\informacion\desktop\exportacion-oracle\node_modules\segfault-handler\src\segfault-handler.cpp (116): segfault_handler
00007FF9395ABD00 (ntdll): (filename not available): RtlInitializeCriticalSection
00007FF93958960B (ntdll): (filename not available): RtlImageNtHeaderEx
00007FF9395E9CBA (ntdll): (filename not available): KiUserExceptionDispatcher
d:\santiagoah\informacion\desktop\exportacion-oracle\node_modules\oracledb\src\njs\src\njsconnection.cpp (5605): Connection::cbDynDefine
d:\santiagoah\informacion\desktop\exportacion-oracle\node_modules\oracledb\src\dpi\src\dpistmtimpl.cpp (757): StmtImpl::defineCallback
00007FF8D7AD0FFE (OraOCIEI12): (filename not available): kpugc
00007FF8D6B173AD (OraOCIEI12): (filename not available): kpcdasg
00007FF8D6AD0BE7 (OraOCIEI12): (filename not available): sigpidu
00007FF8D6AA20DC (OraOCIEI12): (filename not available): sigpidu
00007FF8D84D5D1C (OraOCIEI12): (filename not available): nsrecv
00007FF8D68A6557 (OraOCIEI12): (filename not available): upicui2
00007FF8D7AD9585 (OraOCIEI12): (filename not available): upirtr
00007FF8D7BB5E9E (OraOCIEI12): (filename not available): kpurcsc
00007FF8D7AD2A15 (OraOCIEI12): (filename not available): kpugc
00007FF8D7AD4AE6 (OraOCIEI12): (filename not available): kpugc
00007FF8D793E0CD (OraOCIEI12): (filename not available): OCIStmtFetch2
00007FF8F53C2F0F (OCI): (filename not available): OCIStmtFetch2
d:\santiagoah\informacion\desktop\exportacion-oracle\node_modules\oracledb\src\njs\src\njsconnection.cpp (3367): Connection::DoFetch
d:\santiagoah\informacion\desktop\exportacion-oracle\node_modules\oracledb\src\njs\src\njsresultset.cpp (461): ResultSet::Async_GetRows
00007FF7A3E7C887 (node): (filename not available): uv_fs_poll_getpath
00007FF7A3E6B7A4 (node): (filename not available): uv_once
00007FF7A46B9DCD (node): (filename not available): uncompress
00007FF9390D8364 (KERNEL32): (filename not available): BaseThreadInitThunk
00007FF9395A70D1 (ntdll): (filename not available): RtlUserThreadStart
PID 21448 received SIGSEGV for address: 0x1a9b0229
SymInit: Symbol-SearchPath: '.;d:\santiagoah\informacion\Desktop\exportacion-oracle;C:\Program Files\nodejs;C:\Windows;C:\Windows\system32;SRV*C:\websymbols*http://msdl.microsoft.com/download/symbols;', symOptions: 530, UserName: 'SantiagoAH'
OS-Version: 10.0.14393 () 0x100-0x1
d:\santiagoah\informacion\desktop\exportacion-oracle\node_modules\segfault-handler\src\stackwalker.cpp (941): StackWalker::ShowCallstack
d:\santiagoah\informacion\desktop\exportacion-oracle\node_modules\segfault-handler\src\segfault-handler.cpp (116): segfault_handler
00007FF9395ABD00 (ntdll): (filename not available): RtlInitializeCriticalSection
00007FF93958960B (ntdll): (filename not available): RtlImageNtHeaderEx
00007FF9395E9CBA (ntdll): (filename not available): KiUserExceptionDispatcher
d:\santiagoah\informacion\desktop\exportacion-oracle\node_modules\oracledb\src\njs\src\njsconnection.cpp (5605): Connection::cbDynDefine
d:\santiagoah\informacion\desktop\exportacion-oracle\node_modules\oracledb\src\dpi\src\dpistmtimpl.cpp (757): StmtImpl::defineCallback
00007FF8D7AD0FFE (OraOCIEI12): (filename not available): kpugc
00007FF8D6B173AD (OraOCIEI12): (filename not available): kpcdasg
00007FF8D6AD0BE7 (OraOCIEI12): (filename not available): sigpidu
00007FF8D6AA20DC (OraOCIEI12): (filename not available): sigpidu
00007FF8D84D5D1C (OraOCIEI12): (filename not available): nsrecv
00007FF8D68A6557 (OraOCIEI12): (filename not available): upicui2
00007FF8D7AD9585 (OraOCIEI12): (filename not available): upirtr
00007FF8D7BB5E9E (OraOCIEI12): (filename not available): kpurcsc
00007FF8D7AD2A15 (OraOCIEI12): (filename not available): kpugc
00007FF8D7AD4AE6 (OraOCIEI12): (filename not available): kpugc
00007FF8D793E0CD (OraOCIEI12): (filename not available): OCIStmtFetch2
00007FF8F4EA2F0F (OCI): (filename not available): OCIStmtFetch2
d:\santiagoah\informacion\desktop\exportacion-oracle\node_modules\oracledb\src\njs\src\njsconnection.cpp (3367): Connection::DoFetch
d:\santiagoah\informacion\desktop\exportacion-oracle\node_modules\oracledb\src\njs\src\njsresultset.cpp (461): ResultSet::Async_GetRows
00007FF7A3E7C887 (node): (filename not available): uv_fs_poll_getpath
00007FF7A3E6B7A4 (node): (filename not available): uv_once
00007FF7A46B9DCD (node): (filename not available): uncompress
00007FF9390D8364 (KERNEL32): (filename not available): BaseThreadInitThunk
00007FF9395A70D1 (ntdll): (filename not available): RtlUserThreadStart

Also I've used fetchInfo to explicitly fetch my clob column to string but the same happens.

``` javascript
options.fetchInfo = { "XML": { type: oracledb.STRING } };
```

**Comment by cjbj** [comment]
A testcase would be useful.

Or you can wait until 1.13 is out which has a few fixes in this area; it should be out "soon" (TM).

**Comment by sant123** [comment]
``` javascript
let SegfaultHandler = require("segfault-handler");
SegfaultHandler.registerHandler("crash.log");

let data = require("./oracle");
let oracledb = require("oracledb");

oracledb.outFormat = oracledb.OBJECT;
//oracledb.fetchAsString = [oracledb.CLOB]; // this one fails

oracledb.getConnection(data).then(connection => {
    let query = "SELECT * FROM MYTABLE";
    let options = { resultSet: true };
    options.fetchInfo = { "XML": { type: oracledb.STRING } }; // same above

    connection.execute(query, [], options).then(result => {
        let resultSet = result.resultSet;

        resultSet.getRow(function getRow(err, row) {
            if (err) {
                return console.error(err);
            }
            if (!row) {
                return resultSet.close();
            }

            // Do some stuff with data... then get the next row....
            resultSet.getRow(getRow);
        });
    });
});
```

The last error I got was: FATAL ERROR: v8::ToLocalChecked Empty MaybeLocal.

**Comment by cjbj** [comment]
After hacking it to run, I don't see any crash with 1.12.2.  This only proves again that I need to harsher when I ask for test cases to make sure they are immediately runnable and include sample data!

**Comment by sant123** [comment]
@cjbj please take a look at https://github.com/sant123/testcase-oracledb

**Comment by cjbj** [comment]
@sant123 Nice!  A great example of making life easier (TM).  You'll be glad to know I see the crash with 1.12.2 but don't with 1.13.  So bear with us for a few more days, until we can complete testing on 1.13.

**Comment by sant123** [comment]
Thank you @cjbj ☺ I'll be waiting for the update. I appreciate it.

**Comment by cjbj** [comment]
@Tomkinen @sant123 @yancald A use-after-free problem was fixed in 1.13.  Try it out.

**Comment by sant123** [comment]
Thank you @cjbj!
